### DISCUSSION TITLE
IteratorAssembler: choose which property gets an iteration

### DISCUSSION THREAD

**Discussion by michabbb** [phpro/soap-client#443] [Ideas]
hi @veewee,

when using the "IteratorAssembler" i was wondering why the first property gets the iteration and why this cannot be set by the user (me).

https://github.com/phpro/soap-client/blob/9c616b677df4c7038a17a5c5390becadf3f7b55e/src/Phpro/SoapClient/CodeGenerator/Assembler/IteratorAssembler.php#L39

before I get the well known copy&paste answer 😏  I just want to ask, if you see a problem in choosing the user which property gets the iteration?

because right now it´s:

```php
->addRule(new Rules\TypenameMatchesRule(new Rules\AssembleRule(new Assembler\IteratorAssembler()),
 '/^NameValueListType$/'))
```
but why not:
```php
->addRule(new Rules\TypenameMatchesRule(new Rules\AssembleRule(new Assembler\IteratorAssembler('Value')), 
'/^NameValueListType$/'))
```
so the assembler has an optional parameter which property should be used, because always using the first, is very
random and pure luck if that is the property that really needs to be an array.

thanks.

**Comment by michabbb** [comment]
so to be specific:

```php
class IteratorAssembler implements AssemblerInterface
{
    public function __construct(public ?string $propertyName) {

public function assemble(ContextInterface $context): void
    {
        $class = $context->getClass();

        $propertyToUse = NULL;

        $properties    = $context->getType()->getProperties();
        if ($this->propertyName) {
            foreach ($properties as $property) {
                if ($property->getName() === $this->propertyName) {
                    $propertyToUse = $property;
                    break;
                }
            }
        }
        if (!$propertyToUse) {
            $propertyToUse = count($properties) ? current($properties) : null;
        }
        // the rest of the function, but using $propertyToUse instead of $firstProperty  
    }
}
```

I already created my own assembler, I am only asking to add this simple feature to this package, because I can guess,
I am not the only one who wants to choose, which property gets the iteration.

**Comment by veewee** [comment]
Hello,

The idea behind that assembler is to use it on complex types that wrap a single element with minOccurs=0 and maxOccurs=unbounded. So basically to make the complex type behave like an array instead of an object.

```xml
<complexType name="Categories" mixed="true">
    <sequence>
        <element minOccurs="0" maxOccurs="unbounded" name="Category" type="tns:Category"/>
    </sequence>
</complexType>
```

This will result in

```php
class Categories implements IteratorAggregate
{
    /**
     * @var Category[]
     */
    private $Category;

    public function getIterator()
    {
        return new \ArrayIterator(is_array($this->Category) ? $this->Category : []);
    }
}
```

If your object has multiple properties, it doesn't make sense to make it behave like an iterator.
In that case you could do `$yourObject->getTheProp()` which will be an array and therefore iterable.


Another downside about hardcoding which property to use, is that you won't be able to apply that same assembler on multiple type matches, since the properties could vary. The idea behind the assembler is to generate code if a specific rule matches. A rule could be validated against one specific type, but as well on multiple types. 

Eg by regex: `Array$`
This will make any object that ends on Array an iterable. `CategoryArray`, `ProductArray`, ....
The name of the internal prop might be different for both types matching the single rule.

**Comment by michabbb** [comment]
arg.... i was so optimistic... that this might be a good idea 🙈 but sadly, I have to admit, you are right. i really appreciate you taking the time answering that detailed.  at the end, I guess it´s really easier "not" using return-types. having them makes so many things very complicated 😕

**Reply by veewee** [reply]
No problem, I get it's frustrating.

However in order to fix this type issue properly according to todays standard, I'll need to invest a lot of time of intensive WSDL parsing. Currently my return on investment is around 0. So feel free to consider upping that ROI number by sponsoring :)

**Reply by michabbb** [reply]
@veewee of course! if you tell me, how i can donate ? 😏

**Reply by veewee** [reply]
@michabbb  You can find more information on how to donate on my sponsor page:

https://github.com/sponsors/veewee

**Reply by michabbb** [reply]
done 😋 you should put your "Please consider to donate" in every Readme, because I was unable to find it, I guess like many others. don't be shy asking for donations in every Readme of your projects 😏

**Reply by veewee** [reply]
Thanks, really appreciate it!
Gonna take a look on how I can try to get the wsdl parsing (partially) funded in the near future, so that we can get to better types in the long run.
